src/Type/MixedType.php

```diff
@@ -462,8 +462,13 @@ function () use ($level): string {
 
 	public function toBoolean(): BooleanType
 	{
-		if ($this->subtractedType !== null && StaticTypeFactory::falsey()->equals($this->subtractedType)) {
-			return new ConstantBooleanType(true);
+		if ($this->subtractedType !== null) {
+			if ($this->subtractedType->isSuperTypeOf(new ConstantBooleanType(false))->yes()) {
```

Comment:
This logic is wrong. For example if subtracted type is `bool`, this condition will pass and `toBoolean` will return `true`.

Also many things cast to true or false, making this also wrong.

Here you need something more like `if ($this->subtractedType->toBoolean()->isFalse()->yes())`.

Reply:
great catch!